src/isl.rs

```diff
@@ -0,0 +1,211 @@
+use crate::result::{
+    invalid_schema_error_raw, unresolvable_schema_error, IonSchemaError, IonSchemaResult,
+};
+use crate::system::{SharedTypeStore, TypeId};
+use crate::types::TypeDefinition;
+use ion_rs::value::owned::{OwnedElement, OwnedStruct};
+use ion_rs::value::{Element, Sequence, Struct, SymbolToken};
+use ion_rs::IonType;
+use std::convert::{TryFrom, TryInto};
+
+/// Represents a public facing API for schema constraints [IslConstraint] which stores IslTypeRef
+#[derive(Debug, Clone)]
+pub enum IslConstraint {
+    AllOf(Vec<IslTypeRef>),
+    Type(IslTypeRef),
+}
+
+/// Represents a public facing schema type [IslType] which can be converted to a solid TypeRef using TypeStore
+#[derive(Debug, Clone)]
+pub struct IslType {
+    name: Option<String>,
+    constraints: Vec<IslConstraint>,
+}
+
+impl IslType {
+    pub fn new(name: Option<String>, constraints: Vec<IslConstraint>) -> Self {
```

Comment:
+1

...but we may even want to consider breaking named and anonymous IslTypes into different (Rust) types. (Or should I say different structs?) The reason is because that way we can use the Rust type system to help enforce some schema invariants—for example, a schema cannot have a top-level anonymous type, and a nested/inline type cannot have a name. It seems like that would make it easier for users because they would get a compile time error rather than a runtime error if they use the wrong one.

Just to be clear, this is a "have you considered?" comment, not a "you must" sort of comment.

Reply:
I feel like there needs to be a common representation for all types(whether named or anonymous). Although it maybe required to divide these two types into different structs to enforce some of their specific schema rules. I have created an issue(#25) for this, currently I am going ahead with the suggestion to create different methods for them and later if this is required I can create a common `trait` `TypeDefinition` and its two implementations(structs) for named and anonymous types or even core types.